src/unitquaternion.jl

```diff
@@ -0,0 +1,499 @@
+import Base: +, -, *, /, \, exp, log, ≈, ==, inv, conj
+
+abstract type QuatMap end
+
+"""
+    UnitQuaternion{T,D} <: Rotation
+
+4-parameter attitute representation that is singularity-free. Quaternions with unit norm
+represent a double-cover of SO(3). The `UnitQuaternion` does NOT strictly enforce the unit
+norm constraint, but certain methods will assume you have a unit quaternion. The
+`UnitQuaternion` type is parameterized by the linearization method, which maps quaternions
+to the 3D plane tangent to the 4D unit sphere. Follows the Hamilton convention for quaternions.
+
+There are currently 4 methods supported:
+* `VectorPart` - uses the vector (or imaginary) part of the quaternion
+* `ExponentialMap` - the most common approach, uses the exponential and logarithmic maps
+* `CayleyMap` - or Rodrigues parameters (aka Gibbs vectors).
+* `MRPMap` - or Modified Rodrigues Parameter, is a sterographic projection of the 4D unit sphere
+onto the plane tangent to either the positive or negative real poles.
+
+# Constructors
+```julia
+UnitQuaternion(args...)  # defaults to `CayleyMay`
+UnitQuaternion{T<:Real}(args...)
+UnitQuaternion{D<:QuatMap}(args...)
+UnitQuaternion{T,D}(args...)
+```
+where `args...` can be any of the following:
+- `w,x,y,z` specifying the scalar (real) part `w` and the vector (imaginary) part `x,y,z`
+- `AbstractVector` with elements `[w,x,y,z]` as the first 4 elements
+- `StaticVector{3}` with elements `[x,y,z]` and `w = 0`
+"""
+struct UnitQuaternion{T,D<:QuatMap} <: Rotation{3,T}
+    w::T
+    x::T
+    y::T
+    z::T
+
+    @inline function UnitQuaternion{T,D}(w, x, y, z, normalize::Bool = true) where {T,D}
+        if normalize
+            inorm = inv(sqrt(w*w + x*x + y*y + z*z))
+            new{T,D}(w*inorm, x*inorm, y*inorm, z*inorm)
+        else
+            new{T,D}(w, x, y, z)
+        end
+    end
+
+    UnitQuaternion{T,D}(q::UnitQuaternion) where {T,D} = new{T,D}(q.w, q.x, q.y, q.z)
+end
+
+include("quaternion_maps.jl")
+
+# ~~~~~~~~~~~~~~~ Constructors ~~~~~~~~~~~~~~~ #
+# Use default map
+UnitQuaternion(w::W,x::X,y::Y,z::Z, normalize::Bool = true) where {W,X,Y,Z} =
+    UnitQuaternion{promote_type(W,X,Y,Z),DEFAULT_QMAP}(w,x,y,z, normalize)
```

Comment:
As I wrote in the discussion on `expm`, if you can avoid computing in the type domain it's always neater and the code is more naturally generic.  Here's an efficient way to express this:

```julia
function UnitQuaternion(w,x,y,z, qmap::Type{<:QuatMap}=DEFAULT_QMAP, normalize::Bool = true)
    elements = promote(w,x,y,z)
    UnitQuaternion{eltype(elements), qmap}(elements..., normalize)
end
```

Note that
* We can simply pass the `QuatMap` as a *value*, and allow the constructor to do the annoying work of figuring out how to put that into the type parameters correctly
* We compute the appropriate eltype in the value domain, rather than messing directly with types. It's a natural consequence of using `promote` which would be needed when converting the elements anyway.
* With dispatch you could supply a constructor which takes `normalize` but not `qmap`

Reply:
Awesome. I definitely still have a lot to learn about Julia :) I modified several methods to now compute in value domain, including this constructor.